File: sandbox/plc4go/internal/plc4go/modbus/readwrite/model/ModbusPDUError.go
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
package model

import (
    "encoding/xml"
    "errors"
    "io"
    "plc4x.apache.org/plc4go-modbus-driver/v0/internal/plc4go/spi"
    "plc4x.apache.org/plc4go-modbus-driver/v0/internal/plc4go/utils"
)

// The data-structure of this message
type ModbusPDUError struct {
    ExceptionCode uint8
    ModbusPDU
}

// The corresponding interface
type IModbusPDUError interface {
    IModbusPDU
    Serialize(io utils.WriteBuffer) error
}

// Accessors for discriminator values.
func (m ModbusPDUError) ErrorFlag() bool {
    return true
}

func (m ModbusPDUError) FunctionFlag() uint8 {
    return 0
}

func (m ModbusPDUError) Response() bool {
    return false
}

func (m ModbusPDUError) initialize() spi.Message {
    return m
}

func NewModbusPDUError(exceptionCode uint8) ModbusPDUInitializer {
    return &ModbusPDUError{ExceptionCode: exceptionCode}
}

func CastIModbusPDUError(structType interface{}) IModbusPDUError {
    castFunc := func(typ interface{}) IModbusPDUError {
        if iModbusPDUError, ok := typ.(IModbusPDUError); ok {
            return iModbusPDUError
        }
        return nil
    }
    return castFunc(structType)
}

func CastModbusPDUError(structType interface{}) ModbusPDUError {
    castFunc := func(typ interface{}) ModbusPDUError {
        if sModbusPDUError, ok := typ.(ModbusPDUError); ok {
            return sModbusPDUError
        }
        if sModbusPDUError, ok := typ.(*ModbusPDUError); ok {
            return *sModbusPDUError
        }
        return ModbusPDUError{}
    }
    return castFunc(structType)
}

func (m ModbusPDUError) LengthInBits() uint16 {
    var lengthInBits uint16 = m.ModbusPDU.LengthInBits()

    // Simple field (exceptionCode)
    lengthInBits += 8

    return lengthInBits
}

func (m ModbusPDUError) LengthInBytes() uint16 {
    return m.LengthInBits() / 8
}

func ModbusPDUErrorParse(io *utils.ReadBuffer) (ModbusPDUInitializer, error) {

    // Simple Field (exceptionCode)
    exceptionCode, _exceptionCodeErr := io.ReadUint8(8)
    if _exceptionCodeErr != nil {
        return nil, errors.New("Error parsing 'exceptionCode' field " + _exceptionCodeErr.Error())
    }

    // Create the instance
    return NewModbusPDUError(exceptionCode), nil
}

func (m ModbusPDUError) Serialize(io utils.WriteBuffer) error {
    ser := func() error {

    // Simple Field (exceptionCode)
    exceptionCode := uint8(m.ExceptionCode)
    _exceptionCodeErr := io.WriteUint8(8, (exceptionCode))
    if _exceptionCodeErr != nil {
        return errors.New("Error serializing 'exceptionCode' field " + _exceptionCodeErr.Error())
    }

        return nil
    }
    return ModbusPDUSerialize(io, m.ModbusPDU, CastIModbusPDU(m), ser)
}

func (m *ModbusPDUError) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
    for {
        token, err := d.Token()
        if err != nil {
            if err == io.EOF {
                return nil
            }
            return err
        }
        switch token.(type) {
        case xml.StartElement:
            tok := token.(xml.StartElement)
            switch tok.Name.Local {
            case "exceptionCode":
                var data uint8
                if err := d.DecodeElement(&data, &tok); err != nil {
                    return err
                }
                m.ExceptionCode = data
            }
        }
    }
}

func (m ModbusPDUError) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
    if err := e.EncodeToken(xml.StartElement{Name: start.Name, Attr: []xml.Attr{
            {Name: xml.Name{Local: "className"}, Value: "org.apache.plc4x.java.modbus.readwrite.ModbusPDUError"},
        }}); err != nil {
        return err
    }
    if err := e.EncodeElement(m.ExceptionCode, xml.StartElement{Name: xml.Name{Local: "exceptionCode"}}); err != nil {
        return err
    }
    if err := e.EncodeToken(xml.EndElement{Name: start.Name}); err != nil {
        return err
    }
    return nil
}

File: plc4py/plc4py/protocols/modbus/readwrite/ModbusPDUError.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

from dataclasses import dataclass

from plc4py.api.exceptions.exceptions import PlcRuntimeException
from plc4py.api.exceptions.exceptions import SerializationException
from plc4py.api.messages.PlcMessage import PlcMessage
from plc4py.protocols.modbus.readwrite.ModbusErrorCode import ModbusErrorCode
from plc4py.protocols.modbus.readwrite.ModbusPDU import ModbusPDU
from plc4py.spi.generation.ReadBuffer import ReadBuffer
from plc4py.spi.generation.WriteBuffer import WriteBuffer
from plc4py.utils.ConnectionStringHandling import strtobool
from typing import ClassVar
import math


@dataclass
class ModbusPDUError(ModbusPDU):
    exception_code: ModbusErrorCode
    # Accessors for discriminator values.
    error_flag: ClassVar[bool] = True
    function_flag: ClassVar[int] = 0
    response: ClassVar[bool] = False

    def serialize_modbus_pdu_child(self, write_buffer: WriteBuffer):
        write_buffer.push_context("ModbusPDUError")

        # Simple Field (exceptionCode)
        write_buffer.write_unsigned_byte(
            self.exception_code, logical_name="exceptionCode"
        )

        write_buffer.pop_context("ModbusPDUError")

    def length_in_bytes(self) -> int:
        return int(math.ceil(float(self.length_in_bits() / 8.0)))

    def length_in_bits(self) -> int:
        length_in_bits: int = super().length_in_bits()
        _value: ModbusPDUError = self

        # Simple field (exceptionCode)
        length_in_bits += 8

        return length_in_bits

    @staticmethod
    def static_parse_builder(read_buffer: ReadBuffer, response: bool):
        read_buffer.push_context("ModbusPDUError")

        if isinstance(response, str):
            response = bool(strtobool(response))

        exception_code: ModbusErrorCode = read_buffer.read_enum(
            read_function=ModbusErrorCode,
            bit_length=8,
            logical_name="exception_code",
            response=response,
        )

        read_buffer.pop_context("ModbusPDUError")
        # Create the instance
        return ModbusPDUErrorBuilder(exception_code)

    def equals(self, o: object) -> bool:
        if self == o:
            return True

        if not isinstance(o, ModbusPDUError):
            return False

        that: ModbusPDUError = ModbusPDUError(o)
        return (
            (self.exception_code == that.exception_code)
            and super().equals(that)
            and True
        )

    def hash_code(self) -> int:
        return hash(self)

    def __str__(self) -> str:
        # TODO:- Implement a generic python object to probably json convertor or something.
        return ""


@dataclass
class ModbusPDUErrorBuilder:
    exception_code: ModbusErrorCode

    def build(
        self,
    ) -> ModbusPDUError:
        modbus_pduerror: ModbusPDUError = ModbusPDUError(self.exception_code)
        return modbus_pduerror
